### PR TITLE
Add store gateway lazy expanded posting fuzz test to verify correctness

### PR DESCRIPTION
**Pull request by yeya24**
**What this PR does**:

Added an integration test which utilizes `promqlsmith` as a random query generator to generate random matchers.

The test case will run 2 sets of queriers and store gateways. One with lazy expanded posting enabled and one isn't.
Test case creates a block with predefined series and upload the block to minio used in the integration test.

Then run GetSeries API 1000 times with randomly generated matchers and compare whether the two results are the same or not.

I plan to create another test case in another pr to compare Prometheus and Store Gateway with lazy expanded postings enabled.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by yeya24** [comment]
Tests failed, which is expected. We need to wait for fix https://github.com/thanos-io/thanos/pull/7220

**Comment by alanprot** [comment]
Thanks!

**Comment by harry671003** [comment]
Would the tests catch caching related bugs?

**Comment by yeya24** [comment]
@harry671003
https://github.com/cortexproject/cortex/pull/5833/files#diff-6f602b9f45e5376a283136a62e55152e073076c0deb2505b7575ecce7e7812d2R432

We do run the same query twice and save the results to compare. So it should be able to catch that.

File: integration/query_fuzz_test.go [hunk 1]
```diff
@@ -281,3 +289,188 @@ var comparer = cmp.Comparer(func(x, y model.Value) bool {
...
+		res1, err := c1.Series([]string{matcherStrings}, time.UnixMilli(minT), time.UnixMilli(maxT))
+		require.NoError(t, err)
+		res2, err := c2.Series([]string{matcherStrings}, time.UnixMilli(minT), time.UnixMilli(maxT))
+		require.NoError(t, err)
+
+		// Try again and let requests hit cache.
+		newRes1, err := c1.Series([]string{matcherStrings}, time.UnixMilli(minT), time.UnixMilli(maxT))
+		require.NoError(t, err)
+		newRes2, err := c2.Series([]string{matcherStrings}, time.UnixMilli(minT), time.UnixMilli(maxT))
+		require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (yeya24) [reply integration/query_fuzz_test.go]: Fixed by using different timestamps. Synced offline with @alanprot

File: integration/e2e/util.go [hunk 2]
```diff
@@ -202,3 +211,89 @@ func GetTempDirectory() (string, error) {
...
+		}
+	}()
+
+	app := h.Appender(ctx)
+	for i := 0; i < len(series); i++ {
+
+		var ref storage.SeriesRef
+		start := RandRange(rnd, mint, maxt)
+		for j := 0; j < numSamples; j++ {
+			if ref == 0 {
```
Comment (harry671003) [inline integration/e2e/util.go:249]: nit: Do we need the first if here?

`ref, err = app.Append(ref, series[i], start, float64(i+j))`

**Review by harry671003 — APPROVED** [review]
LGTM, with some nits.

Thanks for this. This should give us confidence is lazy postings optimization.

**Comment by harry671003** [comment]
> I plan to create another test case in another pr to compare Prometheus and Store Gateway with lazy expanded postings enabled.

@yeya24 - Is this possible to do as an end to end test? Is it possible to upload the blocks into Prometheus in e2e? Or do you need a deployment of a Prometheus server remote writing to Cortex?

**Comment by yeya24** [comment]
> @yeya24 - Is this possible to do as an end to end test? Is it possible to upload the blocks into Prometheus in e2e? Or do you need a deployment of a Prometheus server remote writing to Cortex?

I am planning to also generate blocks into Prometheus in the e2e test. Prometheus TSDB can just start from that data dir.

**Review by alanprot — APPROVED** [review]

**Pull request merged**